lib/bundler/definition.rb

```diff
@@ -696,16 +684,24 @@ def converge_dependencies
         end
       end
 
+      changes = false
       # We want to know if all match, but don't want to check all entries
       # This means we need to return false if any dependency doesn't match
       # the lock or doesn't exist in the lock.
-      @dependencies.any? do |dependency|
-        locked_dep = @locked_deps[dependency.name]
-        next true if locked_dep.nil?
+      @dependencies.each do |dependency|
+        unless locked_dep = @locked_deps[dependency.name]
+          changes = true
+          next
+        end
+
+        locked_dep.instance_variable_set(:@type, dependency.type)
```

Comment:
Would it make sense to transfer part of the comment from `fixup_dependency_types` so there is context as to why we call `instance_variable_set`?